client/src/entrypoints/page-chooser.js

```diff
@@ -1,44 +1,85 @@
 /* eslint-disable */
 function createPageChooser(id, pageTypes, openAtParentId, canChooseRoot, userPerms) {
-    var chooserElement = $('#' + id + '-chooser');
-    var pageTitle = chooserElement.find('.title');
-    var input = $('#' + id);
-    var editLink = chooserElement.find('.edit-link');
-
-    $('.action-choose', chooserElement).on('click', function() {
-        var initialUrl = chooserElement.data('chooserUrl');
-        if (openAtParentId) {
-            initialUrl += openAtParentId + '/';
-        }
+  const chooserElement = $('#' + id + '-chooser');
+  const pageTitle = chooserElement.find('.title');
+  const input = $('#' + id);
+  const editLink = chooserElement.find('.edit-link');
+  const chooserBaseUrl = chooserElement.data('chooserUrl');
 
-        var urlParams = {page_type: pageTypes.join(',')};
-        if (canChooseRoot) {
-            urlParams.can_choose_root = 'true';
-        }
-        if (userPerms) {
-            urlParams.user_perms = userPerms;
-        }
+  /*
+  Construct initial state of the chooser from the rendered (static) HTML and arguments passed to
+  createPageChooser. State is either null (= no page chosen) or a dict of id, parentId,
+  adminTitle (the admin display title) and editUrl.
+  The result returned from the page chooser modal (which is ultimately built from the data
+  attributes in wagtailadmin/pages/listing/_page_title_choose.html) is a superset of this, and
+  can therefore be passed directly to chooser.setState.
+  */
+  let state = null;
+  if (input.val()) {
+    state = {
+      id: input.val(),
+      parentId: openAtParentId,
+      adminTitle: pageTitle.text(),
+      editUrl: editLink.attr('href')
+    };
+  }
 
-        ModalWorkflow({
-            url: initialUrl,
-            urlParams: urlParams,
-            onload: PAGE_CHOOSER_MODAL_ONLOAD_HANDLERS,
-            responses: {
-                pageChosen: function(pageData) {
-                    input.val(pageData.id);
-                    openAtParentId = pageData.parentId;
-                    pageTitle.text(pageData.adminTitle);
-                    chooserElement.removeClass('blank');
-                    editLink.attr('href', pageData.editUrl);
-                }
-            }
-        });
-    });
-
-    $('.action-clear', chooserElement).on('click', function() {
+  /* define public API functions for the chooser */
+  const chooser = {
+    getState: () => state,
+    getValue: () => state.id,
+    setState: (newState) => {
+      if (newState) {
+        input.val(newState.id);
+        pageTitle.text(newState.adminTitle);
+        chooserElement.removeClass('blank');
+        editLink.attr('href', newState.editUrl);
+      } else {
         input.val('');
-        openAtParentId = null;
         chooserElement.addClass('blank');
-    });
+      }
+
+      state = newState;
+    },
+
+    openChooserModal: () => {
```

Comment:
As I see it, chooser widgets and chooser modals are two distinct APIs (the rich text editor also uses the latter independently, and wagtail-generic-chooser treats them as separate concepts), and I made a point of leaving the modal API unchanged here. Things that are not using the form widgets should continue to call ModalWorkflow themselves - the presence of the openChooserModal method here doesn't change that.

I agree with your overall point though - we really could do with a higher-level API for modals rather than an ugly boilerplate-y ModalWorkflow() call. I think a good way of doing that would be to turn ModalWorkflow into a base class, and have page-chooser-modal.js etc export a subclass of it (rather than the dict of onload callbacks to pass to ModalWorkflow that they currently 'export'). That'll be an easier task now that we're moving things into /client, and as a bonus it'll stop eslint complaining about ModalWorkflow being capitalised as a class despite not being one...

I like the idea of making it promise-based! In theory ModalWorkflow supports multiple response callbacks, which wouldn't work with promises, but I can't see anything making use of that in practice.